<h2>TFSA-2021-180: Deadlock in mutually recursive <code>tf.function</code> objects</h2>
<h3>CVE Number</h3>
<p>CVE-2021-41213</p>
<h3>Impact</h3>
<p>The <a href="https://github.com/tensorflow/tensorflow/blob/8d72537c6abf5a44103b57b9c2e22c14f5f49698/tensorflow/python/eager/def_function.py#L542">code behind <code>tf.function</code> API</a> can be made to deadlock when two <code>tf.function</code> decorated Python functions are mutually recursive:</p>
<p>```python
import tensorflow as tf</p>
<p>@tf.function()
def fun1(num):
    if num == 1:
        return
    print(num)
    fun2(num-1)</p>
<p>@tf.function()
def fun2(num):
    if num == 0:
        return
    print(num)
    fun1(num-1)</p>
<p>fun1(9)
```</p>
<p>This occurs due to using a non-reentrant <code>Lock</code> Python object.</p>
<p>Loading any model which contains mutually recursive functions is vulnerable. An attacker can cause denial of service by causing users to load such models and calling a recursive <code>tf.function</code>, although this is not a frequent scenario.</p>
<h3>Patches</h3>
<p>We have patched the issue in GitHub commit <a href="https://github.com/tensorflow/tensorflow/commit/afac8158d43691661ad083f6dd9e56f327c1dcb7">afac8158d43691661ad083f6dd9e56f327c1dcb7</a>.</p>
<p>The fix will be included in TensorFlow 2.7.0. We will also cherrypick this commit on TensorFlow 2.6.1, TensorFlow 2.5.2, and TensorFlow 2.4.4, as these are also affected and still in supported range.</p>
<h3>For more information</h3>
<p>Please consult <a href="https://github.com/tensorflow/tensorflow/blob/master/SECURITY.md">our security guide</a> for more information regarding the security model and how to contact us with issues and questions.</p>
<h3>Attribution</h3>
<p>This vulnerability has been reported by members of the Aivul Team from Qihoo 360.</p>